### PR TITLE
[action] evm tx container

### PR DESCRIPTION
**Pull request by dustinxie**
# Description
1. API node receives evm tx, and carry it in a container format
2. defer the type determination (transfer, execution, etc) into workingset. This solves the issue that call to a newly created contract might be classified as transfer (if the API node is not latest height, or the contract is created in the same block)

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/evmtx_container.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	"github.com/pkg/errors"
+)
+
+var (
+	_ EvmTxContainer = (*evmTxContainer)(nil)
+)
+
+type evmTxContainer struct {
+	chainID      uint32
+	evmNetworkID uint32
```
Comment (CoderZhi) [inline action/evmtx_container.go:25]: may not need to be carried by container

File: api/web3server.go [hunk 2]
```diff
@@ -459,33 +459,51 @@ func (svr *web3Handler) sendRawTransaction(in *gjson.Result) (interface{}, error
...
+			return nil, err
+		}
+		if tx.Protected() && tx.ChainId().Uint64() != uint64(cs.EVMNetworkID()) {
+			return nil, errors.Wrapf(errInvalidEvmChainID, "expect chainID = %d, got %d", cs.EVMNetworkID(), tx.ChainId().Uint64())
+		}
+		encoding, sig, pubkey, err := action.ExtractTypeSigPubkey(tx)
+		if err != nil {
+			return nil, err
+		}
+		elp, err := svr.ethTxToEnvelope(tx)
```
Comment (CoderZhi) [inline api/web3server.go:497]: `elp` is the only thing we need to determine inside of workingset

File: api/web3server.go [hunk 3]
```diff
@@ -459,33 +459,51 @@ func (svr *web3Handler) sendRawTransaction(in *gjson.Result) (interface{}, error
...
+			Encoding: iotextypes.Encoding_ETHEREUM_CONTAINER,
+		}
+	} else {
+		tx, err := action.DecodeEtherTx(rawString)
+		if err != nil {
+			return nil, err
+		}
+		if tx.Protected() && tx.ChainId().Uint64() != uint64(cs.EVMNetworkID()) {
+			return nil, errors.Wrapf(errInvalidEvmChainID, "expect chainID = %d, got %d", cs.EVMNetworkID(), tx.ChainId().Uint64())
+		}
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply api/web3server.go:492]: for the `if` part, we don't decode the tx, it was deferred later
and the `else` part will be deleted after hard-fork

File: api/web3server.go (hunk 2 above)
Comment (dustinxie) [reply api/web3server.go:497]: yes, that's right. So what is your suggestion? 
this is in `else` part and will be deleted after hard-fork

File: action/evmtx_container.go (hunk 1 above)
Comment (dustinxie) [reply action/evmtx_container.go:25]: it is needed when calling `Unfold` to convert the tx

File: action/action.go [hunk 4]
```diff
@@ -25,6 +26,11 @@ type (
 		ToEthTx(uint32) (*types.Transaction, error)
 	}
 
+	EvmTxContainer interface {
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply action/action.go:29]: it is for inside workingset, to be able to tell this action is of the container type, and call `Unfold` to convert it

File: action/evmtx_container.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+func (etx *evmTxContainer) hash() hash.Hash256 {
+	h := etx.tx.Hash()
+	return hash.BytesToHash256(h[:])
+}
+
+func (etx *evmTxContainer) To() *common.Address {
+	return etx.tx.To()
+}
+
+func (etx *evmTxContainer) Unfold(isContract bool) (*SealedEnvelope, error) {
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply action/evmtx_container.go:50]: 1. can you explain what's not good using boolean as param of a public func? We can split into `UnfoldContract` and `Unfold`, then 80% code of these 2 would be duplicate

2. the duplicate logic in `web3server.go` will be deleted after hard-fork

File: state/factory/workingset.go [hunk 6]
```diff
@@ -166,6 +166,26 @@ func (ws *workingSet) runAction(
 	if !ok {
 		return nil, errors.New("protocol is empty")
 	}
+	// if it's an EVM tx container, unfold the tx inside
+	if container, ok := selp.Action().(action.EvmTxContainer); ok {
+		var (
+			isContract bool
+			err        error
+		)
+		if container.To() == nil {
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply state/factory/workingset.go:175]: fixed

File: state/factory/workingset.go [hunk 7]
```diff
@@ -166,6 +166,26 @@ func (ws *workingSet) runAction(
...
+			isContract = true
+		} else {
+			ioAddr, _ := address.FromBytes(container.To().Bytes())
+			isContract, err = ws.checkContract(ctx, ioAddr)
+		}
```
Comment (CoderZhi) [inline state/factory/workingset.go:180]: code may be clear as:
`isContract, err := ws.checkContract(ctx, container.To())`

File: state/factory/workingset.go [hunk 8]
```diff
@@ -166,6 +166,26 @@ func (ws *workingSet) runAction(
...
+		if err != nil {
+			return nil, err
+		}
+		selp, err = container.Unfold(isContract)
```
Comment (CoderZhi) [inline state/factory/workingset.go:184]: if this is the only place to call `Unfold`, we may simply the interface to
```
  func Unfold(func (*common.Address) bool)) (*SealedEnvelope, error)
```
and delete `func To() *common.Address`.

File: action/evmtx_container.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (CoderZhi) [reply action/evmtx_container.go:50]: similar to https://github.com/iotexproject/iotex-core/pull/4278/files#diff-2bbca1f3abc131070c6346ef28065660350c41ac7bd106a566ce9b3a7dc0b458R27, Deserializer, the logic of decoding a tx to sealed envelope could be wrapped as a struct
```
type TxConverter struct {
    chainID uint32
    deserializer Deserializer
}

func (etc EVMTxConverter) Convert(tx *types.Transaction) (*SealedEnvelope, error) {
    pbAction := ...
    return etc.deserializer.ActionToSealedEnvelope(pbAction)
}
```
thus, we can unify all the conversion logic, and use the converter in coreservice/workingset/any other places. `chainID` and `evmNetworkID` do not need to be carried by the container.

File: action/sealedenvelope.go [hunk 9]
```diff
@@ -179,6 +197,49 @@ func (sealed *SealedEnvelope) loadProto(pbAct *iotextypes.Action, evmID uint32)
...
+		AbstractAction: AbstractAction{
+			version:  version.ProtocolVersion,
+			chainID:  core.GetChainID(),
+			nonce:    tx.Nonce(),
+			gasLimit: tx.Gas(),
+			gasPrice: tx.GasPrice(),
+		},
+		payload: NewEvmTxContainer(core.GetChainID(), evmID, &tx),
+	}
+	// populdate all fields
```
Comment (CoderZhi) [inline action/sealedenvelope.go:231]: populate

File: api/web3server.go (hunk 2 above)
Comment (CoderZhi) [reply api/web3server.go:497]: To construct an iotextypes.Action, the following fields need to be set:
```
		Core:         ...,
		SenderPubKey: pubkey.Bytes(),
		Signature:    sig,
		Encoding:     encoding,
```
and `Core` is the only place which cannot be determined so far, but we can still fill the other fields.
In this file, which is an API entry point, it doesn't hurt to do https://github.com/iotexproject/iotex-core/pull/4278/files#r1616309346 for the `else` case

File: action/action_deserializer.go [hunk 10]
```diff
@@ -25,7 +25,14 @@ func (ad *Deserializer) SetEvmNetworkID(id uint32) *Deserializer {
...
 func (ad *Deserializer) ActionToSealedEnvelope(pbAct *iotextypes.Action) (*SealedEnvelope, error) {
-	var selp SealedEnvelope
-	err := selp.loadProto(pbAct, ad.evmNetworkID)
+	var (
+		selp SealedEnvelope
+		err  error
+	)
+	if pbAct.GetEncoding() == iotextypes.Encoding_ETHEREUM_CONTAINER {
+		err = selp.loadEvmTransaction(pbAct, ad.evmNetworkID)
+	} else {
```
Comment (CoderZhi) [inline action/action_deserializer.go:34]: unified into `selp.loadProto`

File: action/sealedenvelope.go [hunk 11]
```diff
@@ -179,6 +197,49 @@ func (sealed *SealedEnvelope) loadProto(pbAct *iotextypes.Action, evmID uint32)
...
+	elp := envelope{
+		AbstractAction: AbstractAction{
+			version:  version.ProtocolVersion,
+			chainID:  core.GetChainID(),
+			nonce:    tx.Nonce(),
+			gasLimit: tx.Gas(),
+			gasPrice: tx.GasPrice(),
+		},
+		payload: NewEvmTxContainer(core.GetChainID(), evmID, &tx),
+	}
```
Comment (CoderZhi) [inline action/sealedenvelope.go:230]: unified to `elp.LoadProto`

File: action/sealedenvelope.go [hunk 12]
```diff
@@ -179,6 +197,49 @@ func (sealed *SealedEnvelope) loadProto(pbAct *iotextypes.Action, evmID uint32)
...
+	// populdate all fields
+	sealed.Envelope = &elp
+	sealed.encoding = iotextypes.Encoding_ETHEREUM_CONTAINER
```
Comment (CoderZhi) [inline action/sealedenvelope.go:233]: after converting to a `sealed`, the temp encoding could be replaced

File: state/factory/workingset.go (hunk 8 above)
Comment (dustinxie) [reply state/factory/workingset.go:184]: ok, updated

File: state/factory/workingset.go (hunk 7 above)
Comment (dustinxie) [reply state/factory/workingset.go:180]: updated

File: action/action_deserializer.go (hunk 10 above)
Comment (dustinxie) [reply action/action_deserializer.go:34]: that kind of confused things a little bit conceptually
`loadEvmTransaction` is to temporarily house the evm tx bytes, and make the `Selp.Action()` expose correction funcs (like `Hash(), Cost(), IntrinsicGas()`), so actpool can work correctly 
the actual conversion (which calls `selp.loadProto`) will be called in `Unfold` inside workingset

File: action/sealedenvelope.go (hunk 12 above)
Comment (dustinxie) [reply action/sealedenvelope.go:233]: not yet, here it is only the temporary housing of evmtx bytes, the final conversion to actual `sealed` (with actual action inside) is done in workingset

**Review by CoderZhi — APPROVED** [review]

File: state/factory/workingset.go [hunk 13]
```diff
@@ -166,6 +167,14 @@ func (ws *workingSet) runAction(
 	if !ok {
 		return nil, errors.New("protocol is empty")
 	}
+	// if it's an EVM tx container, unfold the tx inside
+	if protocol.MustGetFeatureCtx(ctx).UseEvmContainer {
+		if container, ok := selp.Action().(action.EvmTxContainer); ok {
+			if err := container.Unfold(selp, ctx, ws.checkContract); err != nil {
+				return nil, errors.Wrapf(err, "Failed to unfold EVM tx inside the container")
```
Comment (CoderZhi) [inline state/factory/workingset.go:174]: if an error is returned for a container with invalid tx, `pickAndRunActions` will quit. Is this an expected behavior?

File: api/web3server.go [hunk 14]
```diff
@@ -477,15 +481,39 @@ func (svr *web3Handler) sendRawTransaction(in *gjson.Result) (interface{}, error
...
+		rawBytes, _ := hex.DecodeString(rawString)
+		req = &iotextypes.Action{
+			Core: &iotextypes.ActionCore{
+				Version:  version.ProtocolVersion,
+				Nonce:    tx.Nonce(),
+				GasLimit: tx.Gas(),
+				GasPrice: tx.GasPrice().String(),
+				ChainID:  cs.ChainID(),
+				Action: &iotextypes.ActionCore_EvmTxContainer{
```
Comment (envestcc) [inline api/web3server.go:496]: seems misleading name, It actually is generic tx container, not only for execution

File: action/evmtx_container.go [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+		return err
+	}
+	selp.Envelope = elp
+	selp.encoding = encoding
+	selp.hash = hash.ZeroHash256
+	selp.srcAddress = nil
+	return nil
+}
+
+func (etx *evmTxContainer) Cost() (*big.Int, error) {
```
Comment (envestcc) [inline action/evmtx_container.go:112]: Is the `Cost` method of other specific action no longer being used?

File: action/evmtx_container.go [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+	maxExecFee := big.NewInt(0).Mul(etx.tx.GasPrice(), big.NewInt(0).SetUint64(etx.tx.Gas()))
+	return maxExecFee.Add(etx.tx.Value(), maxExecFee), nil
+}
+
+func (etx *evmTxContainer) IntrinsicGas() (uint64, error) {
```
Comment (envestcc) [inline action/evmtx_container.go:117]: same question

File: action/sealedenvelope.go [hunk 17]
```diff
@@ -30,6 +30,20 @@ type SealedEnvelope struct {
 // an all-0 return value means the transaction is invalid
 func (sealed *SealedEnvelope) envelopeHash() (hash.Hash256, error) {
 	switch sealed.encoding {
+	case iotextypes.Encoding_ETHEREUM_CONTAINER:
```
Comment (envestcc) [inline action/sealedenvelope.go:33]: prefer Encoding_IOTEX_CONTAINER, b/c It's not format from eth

File: action/evmtx_container.go [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+		return gas, err
+	}
+	if acl := etx.tx.AccessList(); len(acl) > 0 {
+		gas += uint64(len(acl)) * TxAccessListAddressGas
+		gas += uint64(acl.StorageKeys()) * TxAccessListStorageKeyGas
+	}
+	return gas, nil
+}
+
+func (etx *evmTxContainer) SetEnvelopeContext(*AbstractAction) {}
```
Comment (envestcc) [inline action/evmtx_container.go:129]: explicitly panic that the method is not implemented to avoid potential bugs?

File: action/rlp_tx_test.go [hunk 19]
```diff
@@ -499,6 +551,30 @@ func convertToNativeProto(tx *types.Transaction, actType string) *iotextypes.Act
...
+		"transferStake", "candidateRegister", "candidateUpdate", "candidateActivate", "candidateEndorsement", "candidateTransferOwnership":
+		return func(context.Context, *common.Address) (bool, bool, bool, error) {
+			return false, true, false, nil
+		}
+	case "rewardingClaim", "rewardingDeposit":
+		return func(context.Context, *common.Address) (bool, bool, bool, error) {
+			return false, false, true, nil
+		}
+	default:
+		panic("unsupported")
```
Comment (envestcc) [inline action/rlp_tx_test.go:574]: return error instead of panic? the error is never being settled

File: state/factory/workingset.go (hunk 13 above)
Comment (dustinxie) [reply state/factory/workingset.go:174]: 1. it is not expected
2. this applies to other error returns (not a new issue caused by this tx container), added a TODO

File: action/evmtx_container.go (hunk 15 above)
Comment (envestcc) [reply action/evmtx_container.go]: can comment that it's only used for actpool, actual cost is based on the underlying action

**Review by envestcc — APPROVED** [review]
does it require some e2etests?

**Pull request merged**